test/unit/model/DataBaseAccessTest.php

```diff
@@ -172,23 +237,150 @@ public function testGetPermissions($userIds, array $resourceIds)
+                'numEvents' => 0,
+                'numInserts' => 0,
+                'permissionData' => []
+            ],
+            'Empty permissions don\'t result in events nor queries' => [
+                'numEvents' => 0,
+                'numInserts' => 0,
+                'permissionData' => [
+                    [
+                        'resource' => $this->getResourceMock('123'),
+                        'permissions' => []
+                    ]
+                ]
+            ],
+            '3 permissions: One resource, single user' => [
+                'numEvents' => 3,
+                'numInserts' => 1,
+                'permissionData' => [
+                    [
+                        'resource' => $this->getResourceMock('123'),
+                        'permissions' => [
+                            123 => ['GRANT', 'READ', 'WRITE']
+                        ]
+                    ]
+                ]
+            ],
+            '6 permissions: One resource, two users' => [
+                'numEvents' => 6,
+                'numInserts' => 1,
+                'permissionData' => [
+                    [
+                        'resource' => $this->getResourceMock('123'),
+                        'permissions' => [
+                            123 => ['GRANT', 'READ', 'WRITE'],
+                            456 => ['GRANT', 'READ', 'WRITE']
+                        ]
+                    ]
+                ]
+            ]
+        ];
+    }
+
+    public function addMultiplePermissionsBoundaryValues(): Generator
+    {
+        // Should produce 1, 1 & 2 queries
+        $sizes = range(self::INSERT_CHUNK_SIZE - 1, self::INSERT_CHUNK_SIZE + 1);
+
+        $resourceId = 100;
+        foreach ($sizes as $numPermissions) {
+            yield $this->getMultiplePermissionsScenario($resourceId, $numPermissions);
+            $resourceId++;
+        }
+
+        // Should produce 2, 2 & 3 queries
+        $sizes = range((self::INSERT_CHUNK_SIZE * 2) - 1, (self::INSERT_CHUNK_SIZE * 2) + 1);
+
+        $resourceId = 200;
+        foreach ($sizes as $numPermissions) {
+            yield $this->getMultiplePermissionsScenario($resourceId, $numPermissions);
+            $resourceId++;
+        }
+    }
+
+    private function getMultiplePermissionsScenario(int $resourceId, int $numPermissions): array
+    {
+        // A single event and row is generated per user, resource and permission
+
+        return [
+            'numEvents' => $numPermissions,
+            'numInserts' => (int) ceil($numPermissions / self::INSERT_CHUNK_SIZE),
+            'permissionData' => [
+                [
+                    'resource' => $this->getResourceMock((string)($resourceId)),
+                    'permissions' => $this->getPermissionsArrayGenerator($numPermissions)
+                ]
+            ],
+        ];
+    }
+
+    private function getPermissionsArrayGenerator(int $numPermissions): Generator
```

Comment:
That's true, it makes the test harder to write and harder to follow. However, we've already changed the approach by allowing lowering the number of inserted rows per query instead of generating huge arrays of permissions.

This was used as a way to generate big sets of permissions using as less memory as possible for them (instead of building the full array before the test start) in order to test INSERTs were split correctly.